// DB/index.js
var mongo = require('mongoskin');
var db = mongo.db('localhost:4000/ECTABase?auto_reconnect');
var db1 = mongo.db('localhost:4000/test?auto_reconnect');
var DB = {
    IBO: {'AutoPart': '配件', 'AutoArticles': '用品', 'Project': '项目', 'BODefine': '业务对象', 'CombinedObj': '套餐'},
    ICommodity: {'AutoPart': 1, 'AutoArticles': 1, 'Project': 1},
    BODefine: db.collection('BODefine'),
    AutoPart: db1.collection('AutoPart'),
    AutoArticles: db.collection('AutoArticles'),
    Project: db.collection('Project'),
    Price: db.collection('Price'),
    CarTypeClass: db.collection('CarTypeClass'),
    Bill: db1.collection('Bill'),
    CardDiscount: db.collection('CardDiscount'),
    CarTypeSetting: db1.collection('CarTypeSetting'),
    CardType: db.collection('CardType'),
    TierPattern: db1.collection('TierPattern'),
    /**
     * 媒体类{_id:String,ObjType:String,ObjID:String,MediaType:'img/flash/',Url:String,IsMaster:Boolean(主图片)}
     */
    Media: db.collection('Media'),
    Brand: db.collection('Brand'),
    CarExt: db1.collection('CarExt'),
    CombinedObj: db.collection('CombinedObj'),
    Department: db1.collection("Department"),
    Employee: db1.collection('Employee'),
    /**
     * 市场活动
     */
    CRMActivity: db.collection('CRMActivity'),
    /**
     * 预约黑名单 {_id:String,telNum:电话号码}
     */
    ReservationBlackList: db.collection('ReservationBlackList'),
    /**
     * 钣金项目价格
     */
    SheetMetalProjectPrice: db.collection('SheetMetalProjectPrice'),
    /**
     * 钣金项目子项目
     */
    SheetMetalSubProject: db.collection('SheetMetalSubProject')
};


exports.DB = DB;
